### PR TITLE
Made all body, turret and barrel render traits upgradable

### PR DESCRIPTION
**Pull request by reaperrr**
Another milestone towards implementing TS deployed units (artillery, tick tank etc.) via `DeployToUpgrade`.

Allows actors to change appearance on upgrade/levelup too, of course.

### CODE REVIEW

File: mods/ra/weapons/largecaliber.yaml [hunk 1]
```diff
@@ -170,7 +170,7 @@ TurretGun:
 		ValidImpactTypes: Water
 
 8Inch:
-	ReloadDelay: 250
+	ReloadDelay: 50#250
```
Comment (Phrohdoh) [inline mods/ra/weapons/largecaliber.yaml]: Why this change?

Comment (reaperrr) [inline mods/ra/weapons/largecaliber.yaml]: Just for this testcase, to make it upgrade faster.

**Comment by reaperrr** [comment]
Updated, fixed commit message and added `SequenceReferences` to voxel traits.

File: mods/ra/rules/ships.yaml [hunk 2]
```diff
@@ -150,7 +150,7 @@ CA:
 		BuildPaletteOrder: 50
 		Prerequisites: ~syrd, atek, ~techlevel.unrestricted
 	Valued:
-		Cost: 2400
+		Cost: 500#2400
```
Comment (Phrohdoh) [inline mods/ra/rules/ships.yaml]: Why this change?

File: OpenRA.Mods.Common/Traits/Render/WithBarrel.cs [hunk 3]
```diff
@@ -17,24 +17,25 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Renders barrels for units with the Turreted trait.")]
-	class WithBarrelInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>
+	public class WithBarrelInfo : UpgradableTraitInfo, ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<TurretedInfo>,
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithBarrel.cs]: `UpgradableTraitInfo, ITraitInfo` is redundant. Use just `UpgradableTraitInfo`.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Game/Traits/BodyOrientation.cs [hunk 4]
```diff
@@ -62,7 +63,7 @@ public BodyOrientation(ActorInitializer init, BodyOrientationInfo info)
 				if (info.QuantizedFacings >= 0)
 					return info.QuantizedFacings;
 
-				var qboi = self.Info.Traits.GetOrDefault<IQuantizeBodyOrientationInfo>();
+				var qboi = self.Info.Traits.WithInterface<IQuantizeBodyOrientationInfo>().FirstOrDefault();
```
Comment (pchote) [inline OpenRA.Game/Traits/BodyOrientation.cs]: This is bogus for two reasons:

1.  You're unconditionally taking the first implementation of this trait, regardless of whether it is active or not.
2.  Fixing 1. doesn't help, because a bunch of things implicitly depend on `IQuantizeBodyOrientationInfo` not changing.

A better idea might be to move `IQuantizeBodyOrientationInfo` onto `RenderSprites` / `RenderVoxels`, but i'll leave it to you to work out how to make `RenderSprites` work out the number of facings it should use.

File: OpenRA.Mods.Common/Traits/Render/WithBarrel.cs [hunk 5]
```diff
@@ -51,38 +52,43 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
...
-	class WithBarrel
+	public class WithBarrel : UpgradableTrait<WithBarrelInfo>
 	{
-		WithBarrelInfo info;
-		Actor self;
-		Armament armament;
-		Turreted turreted;
-		IBodyOrientation body;
-		Animation anim;
+		public readonly Animation DefaultAnimation;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithBarrel.cs]: This doesn't appear to be used anywhere, so why make it public?

If you need to e.g. run an animation from an external trait then that should be done via an appropriate interface.

File: OpenRA.Mods.Common/Traits/Render/WithVoxelBody.cs [hunk 6]
```diff
@@ -18,16 +18,19 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Also returns a default selection size that is calculated automatically from the voxel dimensions.")]
-	public class WithVoxelBodyInfo : ITraitInfo, IQuantizeBodyOrientationInfo, IRenderActorPreviewVoxelsInfo, Requires<RenderVoxelsInfo>
+	public class WithVoxelBodyInfo : UpgradableTraitInfo, IQuantizeBodyOrientationInfo, IRenderActorPreviewVoxelsInfo, Requires<RenderVoxelsInfo>, Requires<IBodyOrientationInfo>
 	{
-		public readonly string Sequence = "idle";
+		[SequenceReference] public readonly string Sequence = "idle";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithVoxelBody.cs]: These are all voxels, so `[SequenceReference]` is wrong.

File: OpenRA.Game/Traits/BodyOrientation.cs (hunk 4 above)
Comment (penev92) [inline OpenRA.Game/Traits/BodyOrientation.cs]: In what situation would this be a problem and why would `RenderSprites`/`RenderVoxels` deal with the facings?

**Comment by reaperrr** [comment]
Closing this for now.

Refactoring the `IQuantizeBodyOrientation` implementation should be done in its own PR first, as discussed on IRC. 
However, a large portion of traits which currently implement `IQuantizeBodyOrientation` will go away when `RenderSimple`/`RenderBuilding` and their derivatives are replaced by `WithSpriteBody` + XY, so it makes far more sense to do the latter first.

In summary, this PR would likely sit dead for a few more weeks, so there's not point in occupying the PR queue until then.
